src/persist/src/indexed/future.rs

```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    /// Returns the smallest SeqNo greater than all the ones contained.
+    pub fn seqno(&self) -> SeqNo {
+        self.batches.last().map_or(SeqNo(0), |((_, end), _)| *end)
+    }
+
+    /// Writes the given batch to [Blob] storage at the given key and logically
+    /// adds the contained updates to this future.
+    pub fn append<L: Blob>(
+        &mut self,
+        key: String,
+        batch: BlobFutureBatch,
+        blob: &mut BlobCache<L>,
+    ) -> Result<(), Error> {
+        if batch.desc.0 != self.seqno() {
+            return Err(Error::from(format!(
+                "batch lower doesn't match seqno {:?}: {:?}",
+                self.seqno(),
+                batch.desc
+            )));
+        }
+        // TODO: Sort the updates in the batch by `(time, key, value)` (or
+        // ensure that they're sorted, if it turns out this work should have
+        // happened somewhere else).
+        let desc = batch.desc.clone();
+        blob.set_future_batch(key.clone(), batch)?;
+        self.batches.push((desc, key));
+        Ok(())
+    }
+
+    /// Returns a consistent read of the updates contained in this future
+    /// matching the given filters (in practice, everything not in Buffer or
+    /// Trace).
+    pub fn snapshot<L: Blob>(
+        &self,
+        upper: SeqNo,
+        lower: Antichain<u64>,
+        blob: &BlobCache<L>,
+    ) -> Result<FutureSnapshot, Error> {
+        let mut updates = Vec::with_capacity(self.batches.len());
+        for (_, key) in self.batches.iter() {
+            updates.push(blob.get_future_batch(key)?);
+        }
+        Ok(FutureSnapshot {
+            upper,
+            lower,
+            updates,
+        })
+    }
+
+    /// Returns all updates contained in this future in the given time bound.
+    ///
+    /// TODO: Unify this with `snapshot`.
+    pub fn iterate<L: Blob>(
+        &self,
+        lower: &Antichain<u64>,
+        upper: &Antichain<u64>,
+        buf: &mut Vec<((String, String), u64, isize)>,
+        blob: &BlobCache<L>,
+    ) -> Result<(), Error> {
+        for (_, key) in self.batches.iter() {
+            let batch = blob.get_future_batch(key)?;
+            buf.extend(
+                batch
+                    .updates
+                    .iter()
+                    .filter(|(_, _, ts, _)| lower.less_than(ts) && !upper.less_than(ts))
+                    .map(|(_, (key, val), ts, diff)| ((key.clone(), val.clone()), *ts, *diff)),
+            )
+        }
+        Ok(())
+    }
+
+    /// Removes all updates contained in this future before the given upper
+    /// bound.
+    pub fn truncate(&mut self, upper: Antichain<u64>) -> Result<(), Error> {
+        // TODO: Validate that this doesn't regress upper.
+        self.frontier = upper;
+        // TODO: Actually delete the data.
+        Ok(())
+    }
+}
+
+/// A consistent snapshot of the data currently in a persistent [BlobFuture].
+pub struct FutureSnapshot {
+    upper: SeqNo,
```

Comment:
having upper and lower be two different types makes my head hurt :/ I think perhaps `seqno_upper` and `timestamp_lower` could be better names?

Reply:
I almost did this myself the first time. Done!